packages/core/src/2d/assembler/CharAssembler.ts

```diff
@@ -0,0 +1,283 @@
+          }
+        } else {
+          chars += char;
+          charsWidth += charInfo.xAdvance;
+          maxAscent < ascent && (maxAscent = ascent);
+          maxDescent < descent && (maxDescent = descent);
+        }
+      }
+
+      if (charsWidth > 0) {
+        lines.push(chars);
+        lineWidths.push(charsWidth);
+        lineMaxSizes.push({
+          ascent: maxAscent,
+          descent: maxDescent,
+          size: maxAscent + maxDescent
+        });
+      }
+    }
+
+    let height = renderer.height * _pixelsPerUnit;
+    if (renderer.overflowMode === OverflowMode.Overflow) {
+      height = lineHeight * lines.length;
+    }
+
+    return {
+      width,
+      height,
+      lines,
+      lineWidths,
+      lineHeight,
+      lineMaxSizes
+    };
+  }
+
+  private static _measureTextWithoutWrap(renderer: TextRenderer): TextMetrics {
+    const { fontSize, fontStyle } = renderer;
+    const { name } = renderer.font;
+    const fontString = TextUtils.getNativeFontString(name, fontSize, fontStyle);
+    const charFont = renderer._charFont;
+    const fontSizeInfo = TextUtils.measureFont(fontString);
+    const lines = renderer.text.split(/(?:\r\n|\r|\n)/);
+    const lineCount = lines.length;
+    const lineWidths = new Array<number>();
+    const lineMaxSizes = new Array<FontSizeInfo>();
+    const { _pixelsPerUnit } = Engine;
+    const lineHeight = fontSizeInfo.size + renderer.lineSpacing * _pixelsPerUnit;
+    let width = 0;
+    let height = renderer.height * _pixelsPerUnit;
+    if (renderer.overflowMode === OverflowMode.Overflow) {
+      height = lineHeight * lineCount;
+    }
+
+    for (let i = 0; i < lineCount; ++i) {
+      const line = lines[i];
+      let curWidth = 0;
+      let maxAscent = -1;
+      let maxDescent = -1;
+
+      for (let j = 0, m = line.length; j < m; ++j) {
+        const charInfo = CharAssembler._getCharInfo(line[j], fontString, charFont);
+        curWidth += charInfo.xAdvance;
+        const { offsetY } = charInfo;
+        const halfH = charInfo.h * 0.5;
+        const ascent = halfH + offsetY;
+        const descent = halfH - offsetY;
+        maxAscent < ascent && (maxAscent = ascent);
+        maxDescent < descent && (maxDescent = descent);
+      }
+      lineWidths[i] = curWidth;
+      lineMaxSizes[i] = {
+        ascent: maxAscent,
+        descent: maxDescent,
+        size: maxAscent + maxDescent
+      };
+      if (curWidth > width) {
+        width = curWidth;
+      }
+    }
+
+    return {
+      width,
+      height,
+      lines,
+      lineWidths,
+      lineHeight,
+      lineMaxSizes
+    };
+  }
+
+  private static _getCharInfo(char: string, fontString: string, font: Font): CharInfo {
+    let charInfo = font._getCharInfo(char);
+    if (!charInfo) {
+      charInfo = TextUtils.measureChar(char, fontString);
+      font._uploadCharTexture(charInfo, TextUtils.textContext().canvas);
+      font._addCharInfo(char, charInfo);
+    }
+
+    return charInfo;
+  }
+}
```

Comment:
CharAssembler do not have design value now, because you only have one mode of char now!